### PR TITLE
[#316] Upgrade Ruby version

### LINKED ISSUE #316: Upgrade Ruby to 3.1.x
## Why

There is an attempt to upgrade it on https://github.com/nimblehq/rails-templates/pull/312

But there is an issue.

> /.asdf/installs/ruby/3.1.0/lib/ruby/3.1.0/psych/visitors/to_ruby.rb:430:in `visit_Psych_Nodes_Alias': Unknown alias: default (Psych::BadAlias)

After investigation, I found it is from [figaro](https://github.com/laserlemon/figaro) gem.

```
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:60:in `parse'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:56:in `block in raw_configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:56:in `raw_configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:64:in `global_configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:32:in `configuration'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:42:in `each'
from /.asdf/installs/ruby/3.1.0/lib/ruby/gems/3.1.0/gems/figaro-1.2.0/lib/figaro/application.rb:36:in `load'
```

This is because on Ruby 3.1.0, it comes with Psych 4 - https://www.ruby-lang.org/en/news/2021/12/25/ruby-3-1-0-released/

> Psych 4.0 changes Psych.load to use safe_load by default. You may need to use Psych 3.3.2 for migrating to this behavior. [[Bug #17866]](https://bugs.ruby-lang.org/issues/17866)

In the issue, they explained that now YML is loaded with `aliases: false` by default. So when the app tries to load Figaro's `config/application.yml`, it failed because in that file, we're using alias. To fix we need to pass `aliases: true` when [loading the config](https://github.com/laserlemon/figaro/blob/6ae998e8209a08f91a165822abecffc0a0626814/lib/figaro/application.rb#L60) just like they've done on [Rails](https://github.com/rails/rails/commit/179d0a1f474ada02e0030ac3bd062fc653765dbe). 

So in order to upgrade to Ruby 3.1, we need to have this issue fixed on figaro. I'll open the issue. But if no one responds, we will need to fork or look into other alternatives. 
 
## Who Benefits?

developers

### PR DESCRIPTION
**Pull request by andyduong1920**
close #316

## What happened 👀

Upgrade Ruby to 3.2.2

## Insight 📝

- Replace the figaro gem with a fork version ([figjam](https://github.com/laserlemon/figaro/issues/290#issuecomment-1310934776)) that supports the Ruby 3.x ([Figaro issue](https://github.com/laserlemon/figaro/issues/289)) 
- There are other 2 issues that due to the new Ruby version
  - https://github.com/nimblehq/rails-templates/pull/419#discussion_r1250320129 
  - https://github.com/nimblehq/rails-templates/pull/419#discussion_r1250347477

## Proof Of Work 📹

The CI is passed

### CODE REVIEW

File: .template/Gemfile.lock [hunk 1]
```diff
@@ -2,64 +2,70 @@ GEM
   remote: https://rubygems.org/
   specs:
     ast (2.4.2)
-    diff-lcs (1.4.4)
+    diff-lcs (1.5.0)
     docker-api (2.2.0)
       excon (>= 0.47.0)
       multi_json
-    excon (0.90.0)
+    excon (0.100.0)
```
Comment (andyduong1920) [inline .template/Gemfile.lock:9]: Need to upgrade the excon to fix the issue on https://github.com/nimblehq/rails-templates/actions/runs/5441234668/jobs/9894962870

```
An error occurred in a `before(:suite)` hook.
Failure/Error: Docker::Container.get(container_id.strip)
ArgumentError:
  The `:host` parameter should not be set for `unix://` connections.
  When supplying a `unix://` URI, it should start with `unix:/` or `unix:///`.
```

File: .template/addons/docker/Dockerfile.tt [hunk 2]
```diff
@@ -38,7 +38,8 @@ RUN apt-get update -qq && \
 ADD https://dl.yarnpkg.com/debian/pubkey.gpg /tmp/yarn-pubkey.gpg
 RUN apt-key add /tmp/yarn-pubkey.gpg && rm /tmp/yarn-pubkey.gpg && \
     echo "deb http://dl.yarnpkg.com/debian/ stable main" > /etc/apt/sources.list.d/yarn.list && \
-    curl -sL https://deb.nodesource.com/setup_"$NODE_SOURCE_VERSION".x | bash - && \
+    echo 'Package: nodejs\nPin: origin deb.nodesource.com\nPin-Priority: 600' > /etc/apt/preferences.d/nodesource && \
```
Comment (andyduong1920) [inline .template/addons/docker/Dockerfile.tt:41]: As Ruby-3.2.2 slim is based on the Debian-12, and Debian 12 provides a newer Node.js package 18.13.0 from the ﻿deb.debian.org repository, so to override the Node.js version, we have to pin the deb.nodesource.com first ([source](https://github.com/nodesource/distributions/issues/1579))

**Comment by olivierobert** [comment]
So in the end, not moving to using https://github.com/bkeepers/dotenv?

**Comment by malparty** [comment]
> * due to the ne

As this is a forked version with no guarantee for long-term support, I think it might still be preferable to move with DotEnv. In the meantime, this forked version seems preferable over the non-maintained original gem. 🤔

**Comment by andyduong1920** [comment]
> So in the end, not moving to using https://github.com/bkeepers/dotenv?

The Removing Figaro issue is still there @olivierobert https://github.com/nimblehq/rails-templates/issues/394, this PR just solve the #316 they are separate issue, the dotenv replacement seems take longer than the expectation. So to solve the #316 issue, we can go with the fork figaro version first.

**Comment by andyduong1920** [comment]
> > * due to the ne
> 
> As this is a forked version with no guarantee for long-term support, I think it might still be preferable to move with DotEnv. In the meantime, this forked version seems preferable over the non-maintained original gem. 🤔

Yeah I added a comment in the [Remove Figaro issue](https://github.com/nimblehq/rails-templates/issues/394#issuecomment-1617505484), we could replace Figaro with DotENV but I think that is not a small task, we have the draft PR for months already.

Btw, this PR is solving the #316 which is just upgrading the Ruby, this doesn't aims to solve the #394 ;) so there are two separate issues ;)

**Review by malparty — APPROVED** [review]

**Review by hoangmirs — APPROVED** [review]

**Pull request merged**